core/modules/loader/Central.h

```diff
@@ -0,0 +1,133 @@
+#define LSST_QSERV_LOADER_CENTRAL_H_
+
+// system headers
+#include <boost/bind.hpp>
+#include <boost/asio.hpp>
+#include <thread>
+#include <vector>
+
+// Qserv headers
+#include "loader/ClientServer.h"
+#include "loader/MasterServer.h"
+#include "loader/MWorkerList.h"
+#include "loader/WWorkerList.h"
+#include "loader/WorkerServer.h"
+#include "proto/loader.pb.h"
+#include "util/ThreadPool.h"
+
+
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+/// TODO add fileId and row to this so it can be checked in _workerKeyInsertReq
+struct ChunkSubchunk {
+    ChunkSubchunk(int chunk_, int subchunk_) : chunk(chunk_), subchunk(subchunk_) {}
+    int const chunk;
+    int const subchunk;
+    friend std::ostream& operator<<(std::ostream& os, ChunkSubchunk csc);
+};
+
+
+/// This class is 'central' to the execution of the program, and must be around
+/// until the bitter end. As such, it can be accessed by normal pointers.
+/// This class is central to loader workers and the master.
+/// It is the base class for CentralWorker, CentralMaster, and CentralClient
+class Central {
+public:
+    Central(boost::asio::io_service& ioService,
+            std::string const& masterHostName, int masterPort)
+        : _ioService(ioService), _masterAddr(masterHostName, masterPort),
+          _checkDoListThread([this](){ _checkDoList(); }){}
+
+    Central() = delete;
+
+    virtual ~Central();
+
+    void run();
+
+    std::string getMasterHostName() const { return _masterAddr.ip; }
+    int getMasterPort() const { return _masterAddr.port; }
+    NetworkAddress getMasterAddr() { return _masterAddr; }
+
+    uint64_t getNextMsgId() { return _sequence++; }
+
+    int getErrCount() const { return _server->getErrCount(); }
+
+    /// Send the contents of 'sendBuf' to 'host:port'. This waits for the message to be
+    /// sent before returning.
+    void sendBufferTo(std::string const& host, int port, BufferUdp& sendBuf) {
+        _server->sendBufferTo(host, port, sendBuf);
+    }
+
+    // Only allow tracked commands on the queue
+    void queueCmd(util::CommandTracked::Ptr const& cmd) {
+        _queue->queCmd(cmd);
+    }
+
+    bool addDoListItem(DoListItem::Ptr const& item) {
+        return _doList.addItem(item);
+    }
+
+    bool runAndAddDoListItem(DoListItem::Ptr const& item) {
+        _doList.runItemNow(item);
+        return _doList.addItem(item);
+    }
+
+
+    virtual std::string getOurLogId() { return "baseclass"; }
+
+protected:
+    /// Repeatedly check the items on the _doList.
+    void _checkDoList();
+
+    boost::asio::io_service& _ioService;
+
+    /// Initialization order is important.
+    DoList _doList{*this}; ///< List of items to be checked at regular intervals.
+
+    NetworkAddress _masterAddr; ///< Network address of the master node.
+    
+    std::atomic<uint64_t> _sequence{1};
+
+    util::CommandQueue::Ptr _queue{new util::CommandQueue()};
+    util::ThreadPool::Ptr _pool{util::ThreadPool::newThreadPool(10, _queue)};
```

Comment:
it's easy to miss that `_queue` is new'ed in braces above, and then used in the braces init here, and if the order of these 2 lines gets reversed that could lead to crashes. A comment when doing this kind of thing can be helpful.

Reply:
It goes back to the initialization order being important comment. The choice being between construction order being important, but hidden in the base class. Or, having the complexity of multistage construction that is passed to all the child classes and cannot be hidden.